CD4PE.md

```diff
@@ -0,0 +1,327 @@
+
+```
+
+For a full CD4PE environment run the `abs_provision_environment_cd4pe` task.
+For a custom environment, run the individual tasks to provision the desired hosts.
+
+## Set up Gitlab
+The test environment uses a standard installation of Gitlab.
+
+### Install the Gitlab package
+* SSH to the Gitlab host as the root user.
+* Add the GitLab package repository.
+
+```
+curl https://packages.gitlab.com/install/repositories/gitlab/gitlab-ee/script.rpm.sh | sudo bash
+
+
+```
+
+* Install the Gitlab package (replace 'ip-x-x-x-x.amz-dev.puppet.net' with the value for your Gitlab host)
+```
+EXTERNAL_URL="http://ip-x-x-x-x.amz-dev.puppet.net" yum install -y gitlab-ee
+```
+
+### Set the root password
+Once Gitlab is installed and running, navigate to the host in a browser.
+* Set the password for the 'root' account to 'puppetlabs'.
+* Navigate to `admin/application_settings/network` and enable the 'Allow requests to the local network from hooks and services' setting.
+
+### Import the control repo
+In the Gitlab UI:
+* Navigate to New Project → Import Project → Repo by URL
+* Provide a project name and description
+* Provide the control repo URL: https://github.com/puppetlabs/puppetlabs-puppetserver_perf_control.git
+* Click the 'Create project' button
+
+You should now see the control repo in Gitlab.
+* Disable the 'auto-dev-ops' pipeline
+
+### Update the control repo for CD4PE
+CD4PE is designed to use the master branch as the default with additional branches per environment.
+
+* Create a 'master' branch from the 'production' branch.
+* Set the 'master' branch as the default.
+* Create additional branches per environment.
+* Remove extraneous branches.
+* Ensure all branches are unprotected.
+
+# Set up PE
+## Enable Code Manager
+We'll be using code manager to deploy the cd4pe module, so we'll set it up before proceeding based on [the docs](https://puppet.com/docs/pe/2019.0/code_mgr_config.html#enable-code-manager-after-installation)
+
+### Prepare the master
+@TODO: automate
+* SSH to the master
+* Add a entry for the Gitlab host to `/etc/hosts`:
+```
+<GITLAB_HOST_IP_ADDRESS>	gplt-gitlab
+```
+
+* Add the following to `~/.ssh/config`:
+```
+Host *
+  StrictHostKeyChecking no
+  UserKnownHostsFile /dev/null
+Host gplt-gitlab
+  User git
+  IdentityFile /etc/puppetlabs/puppetserver/ssh/id-control_repo.rsa
+  PreferredAuthentications publickey
+  UserKnownHostsFile=/dev/null  
+```
+
+* Prepare directory for control repo key
+```
+cd /etc/puppetlabs/puppetserver && mkdir ssh && chown pe-puppet:pe-puppet ssh
+```
+
+* Generate the key
+```
+ssh-keygen -t rsa -b 2048 -C "gplt@puppet.com" -f /etc/puppetlabs/puppetserver/ssh/id-control_repo.rsa -q -N "" 
+
+```
+
+* Set file permissions
+```
+cd ssh && chown pe-puppet:pe-puppet id-control_repo.rsa id-control_repo.rsa.pub
+
+```
+
+* Get the public key
+```
+cat id-control_repo.rsa.pub
+```
+
+* Copy the key and add it to Gitlab
```

Comment:
could use a touch more detail on where in gitlab to add it.  I can't remember at the moment, but is it repo specific or general to gitlab... that kind of thing.

Reply:
Updated